packages/react-server/src/flight/ReactFlightAsyncDispatcher.js

```diff
@@ -25,17 +31,42 @@ function resolveCache(): Map<Function, mixed> {
   return new Map();
 }
 
+function getActiveCache(): AsyncCache {
+  const outerCache: AsyncCache | null =
+    previousAsyncDispatcher !== null
+      ? previousAsyncDispatcher.getActiveCache()
+      : null;
+
+  const innerCache = resolveCache();
+
+  if (outerCache === null) {
+    return innerCache;
+  }
+
+  // If both caches are active, reads will prefer the outer cache
+  // Writes will go into both caches.
+  const chainedCache: AsyncCache = {
```

Comment:
This is a hot path so we shouldn't allocate in it. Instead we should be able to structure it so that we don't need to.

For example if this is the logic we want, then a better approach might be to just have to methods on the dispatcher. One for get and one for set.

However, I wonder if it's not better to just delegate to one cache over another.

It's not also not good that we don't preserve hidden classes here. It should either always be a WeakMap or never a WeakMap. For example if you load a graph first and then do a bunch of renders and then later load react-markup you'd get a bunch of hidden class breakages causing recompilation but also it'd not assume that it might be either WeakMap or this other thing so all lookups would be megamorphic instead of just checking the one.

Reply:
What if when we start a new request, we check if we're already inside someone else's render (the dispatcher returns a cache) and if so we just use their cache?

It seems reasonable that an inner render shares the cache of an outer render.

Reply:
I had the same concerns but couldn't see a way to make it work when chaining at module evaluation time. Deferring to the existing cache if it exists seems doable (ignoring the ownerstack for now which I need to reconcile still): [`2ecf3ec` (#30736)](https://github.com/facebook/react/pull/30736/commits/2ecf3ec04a4bf9d3b11eada6c720cf4e78db28e9)